File: gcode/src/main/java/org/qw3rtrun/p3d/g/decoder/CapabilityReportDecoder.java
package org.qw3rtrun.p3d.g.decoder;

import org.apache.commons.lang3.StringUtils;
import org.qw3rtrun.p3d.g.event.CapabilityReport;
import org.qw3rtrun.p3d.g.event.CapabilityReportEvent;

import java.util.Optional;

public class CapabilityReportDecoder implements GEventDecoder<CapabilityReportEvent> {

    private static String CAP_PREFIX = "Cap";

    private static String SEPARATOR = ":";

    @Override
    public Optional<CapabilityReportEvent> decode(String line) {
        if (line.length() < 7 || !StringUtils.startsWithIgnoreCase(line, CAP_PREFIX)) {
            return Optional.empty();
        }
        var parts = line.split(SEPARATOR);
        if (parts.length != 3) {
            return Optional.empty();
        }

        var capability = parts[1];
        var enabled = parts[2];
        return StringUtils.isNotBlank(capability) ?
                Optional.of(new CapabilityReport(line, capability, parseEnabled(enabled)))
                : Optional.empty();
    }

    boolean parseEnabled(String enabled) {
        if (StringUtils.isNumeric(enabled)) {
            return Integer.parseInt(enabled) > 0;
        }
        return "true".equalsIgnoreCase(enabled);
    }
}

File: gcode/src/main/kotlin/org/qw3rtrun/p3d/g/decoder/CapabilityReportDecoder.kt
package org.qw3rtrun.p3d.g.decoder

import org.apache.commons.lang3.StringUtils
import org.qw3rtrun.p3d.core.msg.CapabilityReport
import org.qw3rtrun.p3d.core.msg.CapabilityReportEvent
import java.util.Optional

class CapabilityReportDecoder : GEventDecoder<CapabilityReportEvent> {

    override fun decode(line: String): Optional<CapabilityReportEvent> {
        if (line.length < 7 || !line.startsWith(CAP_PREFIX, ignoreCase = true)) {
            return Optional.empty()
        }
        val parts = line.split(SEPARATOR)
        if (parts.size != 3) {
            return Optional.empty()
        }

        val capability = parts[1]
        val enabled = parts[2]
        return if (StringUtils.isNotBlank(capability)) {
            Optional.of(CapabilityReport(line, capability, parseEnabled(enabled)))
        } else {
            Optional.empty()
        }
    }

    internal fun parseEnabled(enabled: String): Boolean {
        if (StringUtils.isNumeric(enabled)) {
            return enabled.toInt() > 0
        }
        return "true".equals(enabled, ignoreCase = true)
    }

    companion object {
        private const val CAP_PREFIX = "Cap"
        private const val SEPARATOR = ":"
    }
}
